packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,54 @@
+import { cloneDeep } from 'lodash'
+import { z } from 'zod'
+
+/**
+ * Easily validate vue props, improve error messages drastically
+ *
+ * This augments/replaces Vue’s prop.validator approach by using zod and therefore actually
+ * prints errors to the console that can actually help developers rather than just say
+ * "hey you failed the validator, good luck!" as vue would do by default.
+ *
+ * @example
+ * sections: {
+ * 	required: true,
+ * 	type: Array,
+ * 	validator: propValidator(KottiUserMenu.propsInternalSchema, 'sections'),
+ * }
+ */
+export const propValidator = <
+	SCHEMA extends z.ZodObject<z.ZodRawShape>,
+	KEY extends keyof z.infer<SCHEMA>,
+>(
+	schema: SCHEMA,
+	key: KEY,
+) => {
+	const validator = (value: unknown) => {
+		const result = schema.pick({ [key]: true }).safeParse({ [key]: value })
+
+		if (result.success) return true
+
+		/* eslint-disable no-console */
+		console.group(`propValidator found issues with prop “${key}”`)
+
+		// HACK: 'error' in result is necessary as `ts-jest` doesn’t see that result.success was already properly checked to be falsy and that error now exists
+		if ('error' in result) console.error(result.error)
```

Comment:
It's a stylistic preference

but now when I have those if-conditions particularly for ts not seeing proper types because they are not asserted and are evaluated at runtime

what I do is:

`console.error((result as Exclude<...>).error)`

if you can achieve a similar type asserion, then fix the type of the result with an assertion right below the check for success

~
ok apparently if you type assert, eslint/ts will remove what it redeems as an unnecessary assertion according to the compiler